src/braket/circuits/pauli_string.py

```diff
@@ -0,0 +1,151 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+from typing import List, Tuple, Union
+
+from braket.circuits import Circuit
+
+_IDENTITY = "I"
+_PAULI_X = "X"
+_PAULI_Y = "Y"
+_PAULI_Z = "Z"
+_PAULI_INDICES = {_IDENTITY: 0, _PAULI_X: 1, _PAULI_Y: 2, _PAULI_Z: 3}
+_SIGN_MAP = {"+": 1, "-": -1}
+
+
+class PauliString:
+    """
+    A lightweight representation of a Pauli string with its phase.
+    """
+
+    def __init__(self, pauli_string: Union[str, PauliString]):
+        """
+        Args:
+            pauli_string (Union[str, PauliString]): The representation of the pauli word, either a
+                string or another PauliString object. A valid string consists of an optional phase,
+                specified by an optional sign +/- and optional imaginary unit (both i and j are
+                acceptable) followed by an uppercase string in {I, X, Y, Z}. Example valid strings
+                are: XYZ, +YIZY, -YX, iXXY, -jZYX
+        """
+        if not pauli_string:
+            raise ValueError("pauli_string must not be empty")
+        if isinstance(pauli_string, PauliString):
+            self._phase = pauli_string._phase
+            self._str = pauli_string._str
+            self._indices = pauli_string._indices
+            self._eigenstate_circuits = pauli_string._eigenstate_circuits
+        elif isinstance(pauli_string, str):
+            self._phase, factors_str = PauliString._split(pauli_string)
+            self._str = f"{PauliString._phase_to_str(self._phase)}{factors_str}"
+            self._indices = [_PAULI_INDICES[letter] for letter in factors_str]
+            self._eigenstate_circuits = {}
+        else:
+            raise TypeError(f"Pauli word {pauli_string} must be of type {PauliString} or {str}")
+
+    @property
+    def phase(self) -> complex:
+        """int: The phase of the Pauli string.
+
+        Can be one of {1, -1, 1j, -1j}
+        """
+        return self._phase
+
+    def eigenstate(self, signs: Union[str, List[int], Tuple[int, ...]]) -> Circuit:
```

Comment:
maybe we could default to `signs` being all "+" states rather than having to write `PauliString("XY").eigenstate("+"*len("XY"))` or so.